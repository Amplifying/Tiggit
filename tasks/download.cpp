#include "download.hpp"
#include "curl.hpp"

using namespace Tasks;

std::string DownloadTask::userAgent = "Tiggit/1.0 - see http://tiggit.net/";

struct DLProgress : cURL::Progress
{
  Jobify::JobInfoPtr info;

  bool progress(int64_t total, int64_t now)
  {
    info->current = now;
    info->total = total;

    // Abort the download if the user requested it.
    if(info->checkStatus())
      return false;

    return true;
  }
};

void DownloadTask::doJob()
{
  DLProgress prog;
  prog.info = info;

  if(stream)
    {
      setBusy("Downloading " + url);
      cURL::get(url, stream, userAgent, &prog);
    }
  else
    {
      setBusy("Downloading " + url + " to " + file);
      cURL::get(url, file, userAgent, &prog);
    }

  // All error handling is done through exceptions. If we get here,
  // everything is OK.
  setDone();
};
