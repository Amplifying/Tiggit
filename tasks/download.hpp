#ifndef __TASKS_DOWNLOAD_HPP_
#define __TASKS_DOWNLOAD_HPP_

#include <job/job.hpp>
#include <mangle/stream/stream.hpp>

namespace Tasks
{
  struct DownloadTask : Jobify::Job
  {
    DownloadTask(const std::string &_url, const std::string &_file,
                 Jobify::JobInfoPtr info)
      : Jobify::Job(info), url(_url), file(_file) {}

    DownloadTask(const std::string &_url, Mangle::Stream::StreamPtr _stream,
                 Jobify::JobInfoPtr info)
      : Jobify::Job(info), url(_url), stream(_stream) {}

    static std::string userAgent;

  private:
    void doJob();

    std::string url, file;
    Mangle::Stream::StreamPtr stream;
  };
}

#endif
