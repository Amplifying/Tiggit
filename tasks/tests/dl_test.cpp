#include "../download.hpp"

#include <iostream>
using namespace std;

int main()
{
  string url = "http://tiggit.net/client/latest.tig";
  cout << "Downloading " << url << "...\n";

  Jobify::JobInfoPtr info(new Jobify::JobInfo);
  Tasks::DownloadTask dlj(url, "_output.txt", info);

  dlj.run();

  if(info->isSuccess())
    cout << "Success!\n";
  else
    cout << "Failure: " << info->message << endl;

  return 0;
}
