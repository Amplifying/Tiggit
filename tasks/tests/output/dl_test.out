Downloading http://tiggit.net/client/latest.tig...
Success!
